tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -1802,51 +1804,94 @@ class LocationSet {
-  int useNextLocs(uint32_t count, uint32_t index = 0) {
-    assert(index < kMaxIndex);
-    auto &locs = usedLocs[index];
-    auto &next = nextLoc[index];
-    while (locs[next])
-      next++;
+    auto &set = usedLocations[index];
+    if (loc >= set.size()) {
+      set.resize(std::max<size_t>(loc + 1, set.size() * 2));
+    }
+    set.set(loc);
 
-    int toUse = next;
+    // Maybe the bit marked is the next available one.
+    // Otherwise, there is a hole, ignoring.
+    if (loc == nextAvailableLocation[index]) {
+      ++nextAvailableLocation[index];
+    } else {
+      nextAvailableLocation[index] =
+          std::max(loc + 1, nextAvailableLocation[index]);
+    }
+  }
 
-    for (uint32_t i = 0; i < count; ++i) {
-      assert(!locs[next]);
-      locs.set(next++);
+  // Find the first range of size |count| of unused locations,
+  // and marks them as used.
+  // Returns the first index of this range.
+  int useNextLocs(uint32_t count, uint32_t index = 0) {
+    auto res = find_unused_range(index, count);
+    auto &locations = usedLocations[index];
+
+    // Simple case: no hole large enough left, resizing.
+    if (res == std::nullopt) {
+      const uint32_t spaceLeft =
+          locations.size() - nextAvailableLocation[index];
+      assert(spaceLeft < count && "There is a bug.");
+      const uint32_t requiredAlloc = count - spaceLeft;
+      ;
+      locations.resize(locations.size() + requiredAlloc);
+      res = nextAvailableLocation[index];
     }
 
-    return toUse;
+    for (uint32_t i = res.value(); i < res.value() + count; i++) {
+      locations.set(i);
+    }
+    nextAvailableLocation[index] =
+        std::max(res.value() + count, nextAvailableLocation[index]);
+    return res.value();
   }
 
   /// Returns true if the given location number is already used.
   bool isLocUsed(uint32_t loc, uint32_t index = 0) {
     assert(index < kMaxIndex);
-    return usedLocs[index][loc];
+    if (loc >= usedLocations[index].size())
+      return false;
+    return usedLocations[index][loc];
   }
 
 private:
-  llvm::SmallBitVector usedLocs[kMaxIndex]; ///< All previously used locations
-  uint32_t nextLoc[kMaxIndex];              ///< Next available location
+  // Find the first unused range of size |size| in the given set.
+  // If the set contains such range, returns the first usable index.
+  // Otherwise, nullopt is returned.
+  std::optional<uint32_t> find_unused_range(uint32_t index, uint32_t size) {
+    assert(index < kMaxIndex);
+    const auto &locations = usedLocations[index];
+    uint32_t slot_size = 0;
+    for (uint32_t i = nextAvailableLocation[index]; i < locations.size(); i++) {
+      if (locations[i])
+        slot_size = 0;
+      else
+        ++slot_size;
+
+      if (slot_size == size)
+        return i + 1 - slot_size;
+    }
+
+    return std::nullopt;
+  }
+
+  /// All previously used locations
+  llvm::SmallBitVector usedLocations[kMaxIndex];
+  /// Next available location
```

Comment:
What does "available" mean? Is it the first unset bit in the bitvector? It is one past the last bit that is set?

Reply:
It might be useful to make the comment more precise.

Reply:
Rewrote this class and expanded comments to make this more clear